src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,117 @@ public SubstitutionMatrix(final byte[] matrix) {
-                codeByBase[refBase][baseByCode[refBase][code]] = code;
+                codeByBase[refBase.getBase()][baseByCode[refBase.getBase()][code]] = code;
             }
         }
     }
 
+    /**
+     * Given a reference base and a read base, find the corresponding substitution code
+     * @param refBase reference base being substituted
+     * @param readBase read base to substitute for the reference base
+     * @return code to be used for this refBase/readBase pair
+     */
+    public byte code(final byte refBase, final byte readBase) {
+        if (Character.isLowerCase((char) refBase)) {
+            log.warn(String.format("CRAM: Attempt to generate a substitution code for a lower case reference base '%c'", (char) refBase));
+        }
+        return codeByBase[refBase][readBase];
+    }
+
+    /**
+     * Given a reference base and a substitution code, return the corresponding substitution base.
+     * @param refBase reference base being substituted
+     * @param code substitution code
+     * @return base to be substituted for this (refBase, code) pair
+     */
+    public byte base(final byte refBase, final byte code) {
+        final byte base = baseByCode[refBase][code];
+        if (base == NO_BASE) {
+            // attempt to retrieve a code for a reference base that isn't in the substitution matrix
+            throw new IllegalArgumentException(String.format("CRAM: Attempt to retrieve a substitution base for '%c'", (char) refBase));
+        }
+        return base;
+    }
+
     /**
      * Return this substitution matrix as a byte array in a form suitable for serialization.
-     * @return
+     * @return the encoded matrix in serialized form
      */
     public byte[] getEncodedMatrix() {
         return encodedMatrixBytes;
     }
 
+    @Override
+    public String toString() {
+        final StringBuilder stringBuilder = new StringBuilder();
+        for (final SubstitutionBase r : SubstitutionBase.values()) {
+            stringBuilder.append((char) r.getBase());
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[r.getBase()][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        // lower case substitutions
+        for (final SubstitutionBase r : SubstitutionBase.values()) {
+            stringBuilder.append(Character.toLowerCase((char)r.getBase()));
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[Character.toLowerCase((char)r.getBase())][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        return stringBuilder.toString();
+    }
+
+    // Initialize the baseByCode array to a sentinel value
+    private void initializeBaseByCode() {
+        for (final byte[] base : baseByCode) {
+            Arrays.fill(base, NO_BASE);
```

Comment:
NO_BASE is 0 so all this filling actually does nothing because those arrays are initialized with zero by default. If this method is called somewhere else apart from the beginning of the constructor code then it should be call ```clearBaseByCode()``` instead.

Reply:
Right - this used to use a non-zero value, not now its redundant. Removed.